### ISSUE TITLE
Conflicting results when drawing samples from a trained variational model in different ways

### ISSUE THREAD

**Issue by smsharma** [cornellius-gp/gpytorch#1313]
https://github.com/cornellius-gp/gpytorch/blob/1b97d65f1f9a73e22ecf1792fe17a98dd5a0f04b/gpytorch/models/pyro/_pyro_mixin.py#L18

Hello, I'm somewhat confused about the purpose of the line referenced above when using Approximate GPs with Pyro integration. My impression was that the line just before it,
https://github.com/cornellius-gp/gpytorch/blob/1b97d65f1f9a73e22ecf1792fe17a98dd5a0f04b/gpytorch/models/pyro/_pyro_mixin.py#L17
draws the variational samples. I'm then confused how the full covariance is discarded, and only the diagonal covariance (`function_dist.stddev`) is used to draw the sample.

A consequence is that I get different predictive samples if evaluating a model directly (which is equivalent to the `function_dist = self(input)` line), and if using `pyro.infer.Predictive` (which I believe calls the guide, so equivalent to using the function referenced above). The latter is more "jagged" since it doesn't use covariance information between the data points.

An example demonstrating this with a slightly modified version of [an example in the GPyTorch documentation](https://docs.gpytorch.ai/en/latest/examples/07_Pyro_Integration/Pyro_GPyTorch_Low_Level.html) is given [here]( 
https://nbviewer.jupyter.org/gist/smsharma/96649830b7f13657472599ca9142a196).

I was wondering which would be the "correct" way to draw samples from a trained model, and wanted to clarify my understanding of the line above (I am probably missing something basic here). Thank you very much!

**Comment by gpleiss** [comment]
Hmm this is a good point. I think `pyro.infer.Predictive` is a good way to draw samples - the code in `_pyro_mixin` is designed for variational optimization (i.e. computing the ELBO), which factorizes over data points and thus throws away covariance information. You are right though: if your goal is to draw samples from the approximate predictive posterior, throwing away this covariance information is undesirable.

I think it would make sense to make `function_dist` behave differently in `train` and `eval` mode - where `train` is for computing the ELBO (throw away covariance information) and `eval` model is for computing the approximate predictive posterior (keep covariance information). @jacobrgardner @martinjankowiak thoughts?

**Comment by jacobrgardner** [comment]
@gpleiss Yeah I think having it behave differently in train and eval mode is 100% the way. Case in point, our old rusty `UnwhitenedVariationalStrategy` does things this way:
https://github.com/cornellius-gp/gpytorch/blob/fc59d9a1fe0a20b7ad60f7a4d0f64a1d6e6bf6ea/gpytorch/variational/unwhitened_variational_strategy.py#L168-L178
